state/state.go

```diff
@@ -711,6 +711,14 @@ func (st *State) SetModelAgentVersion(newVersion version.Number, ignoreAgentVers
 			return nil, jujutxn.ErrNoOperations
 		}
 
+		currentNumber, err := version.Parse(currentVersion)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if currentNumber.Compare(newVersion) == 1 {
+			return nil, errors.Errorf("target version %s is less than the current version %s", newVersion, currentVersion)
+		}
```

Comment:
I don't think the intent is to be that strict. Specifically, we could allow rolling back to an earlier patch level.

The core of the original bug was that "juju upgrade-juju" allowed passing in things that the internal workers would reject.

Reply:
The internal workers would reject any downgrade - `state.checkUpgradeInfoSanity` https://github.com/juju/juju/blob/develop/state/upgrade.go#L532 (which is where the error in the bug is coming from) won't allow the version change unless the previous version is strictly less than the target version (unless I'm misreading it).

```go
	if previousVersion.Compare(targetVersion) != -1 {
		return nil, errors.Errorf("cannot sanely upgrade from %s to %s", previousVersion, targetVersion)
	}
```

I talked with @wallyworld and @howbazaar about this (since you were asleep :). They both agreed that we couldn't downgrade because we don't have inverses for the upgrade steps, so there might have been a schema change or something else that we now can't undo. What do you think?